core/templates/expressions/expression-syntax-tree.service.ts

```diff
@@ -16,417 +16,447 @@
         ' does not match expected type ' + this.expectedType;
-      }
-      return this.name + ': ' + this.arg + ' has type ' + this.actualType +
+  }
+  return this.name + ': ' + this.arg + ' has type ' + this.actualType +
         ' which does not match expected type ' + this.expectedType;
-    };
+};
 
-    var getParamsUsedInExpression = function(expression) {
-      var _findParams = function(parseTree) {
-        var paramsFound = [];
+@Injectable({
+  providedIn: 'root'
+})
+export class ExpressionSyntaxTreeService {
+  constructor(private expressionParserService: ExpressionParserService) {}
 
-        if (parseTree instanceof Array) {
-          if (parseTree[0] === '#') {
-            paramsFound.push(parseTree[1]);
-          } else {
-            for (var i = 1; i < parseTree.length; i++) {
-              paramsFound = paramsFound.concat(_findParams(parseTree[i]));
-            }
+  public getParamsUsedInExpression(expression: string): Array<string> {
+    let _findParams = (parseTree: Array<string>|string): Array<string> => {
+      let paramsFound = [];
+      if (parseTree instanceof Array) {
+        if (parseTree[0] === '#') {
+          paramsFound.push(parseTree[1]);
+        } else {
+          for (let i = 1; i < parseTree.length; i++) {
+            paramsFound = paramsFound.concat(_findParams(parseTree[i]));
           }
         }
+      }
 
-        var uniqueParams = [];
-        for (var i = 0; i < paramsFound.length; i++) {
-          if (uniqueParams.indexOf(paramsFound[i]) === -1) {
-            uniqueParams.push(paramsFound[i]);
-          }
+      let uniqueParams = [];
+      for (let i = 0; i < paramsFound.length; i++) {
+        if (uniqueParams.indexOf(paramsFound[i]) === -1) {
+          uniqueParams.push(paramsFound[i]);
         }
+      }
 
-        return uniqueParams.sort();
-      };
-
-      var parsed = ExpressionParserService.parse(expression);
-      return _findParams(parsed);
+      return uniqueParams.sort();
     };
 
-    // Checks if the args array has the expectedNum number of elements and
-    // throws an error if not. If optional expectedMax is specified, it
-    // verifies the number of args is in [expectedNum, expectedMax] range
-    // inclusive.
-    var verifyNumArgs = function(args, expectedNum, expectedMax = expectedNum) {
-      if (expectedMax === undefined) {
-        expectedMax = expectedNum;
-      }
-      if (args.length >= expectedNum && args.length <= expectedMax) {
-        return;
-      }
-      throw new ExprWrongNumArgsError(args, expectedNum, expectedMax);
-    };
+    let parsed = this.expressionParserService.parse(expression);
 
-    var _verifyArgTypesMatchExpectedType = function(argTypes, expectedType) {
-      for (var i = 0; i < argTypes.length; i++) {
-        if (argTypes[i] !== expectedType) {
-          throw new ExprWrongArgTypeError(null, argTypes[i], expectedType);
-        }
-      }
-      return true;
-    };
+    return _findParams(parsed);
```

Comment:
Since we're only using this function once, let's just extract the body and get rid of it.

Reply:
Ah didn't catch that. Let's extract it out and make it a private method instead, then?

Reply:
@brianrodri Alright 👍